### PR TITLE
Add str.substr() and str.substr_until() methods

### PR DESCRIPTION
**Pull request by Xinayder**
`str.substr(start_index: usize)` returns a substring of str starting from the specified point `(Hello.substr(1) = ello)`

`str.substr_until(start_index: usize, length: usize)` returns a substring of str starting from the specified point, with the specified length `(Hello.substr_until(0, 2) = He)`

### CODE REVIEW

File: src/libcore/str/mod.rs [hunk 1]
```diff
@@ -1905,6 +1911,18 @@ impl StrExt for str {
 
     #[inline]
     fn parse<T: FromStr>(&self) -> Result<T, T::Err> { FromStr::from_str(self) }
+    
+    #[inline]
+    fn substr_until(&self, start_index: usize, length: usize) -> Option<&str> {
+        if length == 0 || self.is_empty() { return None; }
+        if start_index == 0 && length == self.len() { return Some(self); }
```
Comment (oli-obk) [inline src/libcore/str/mod.rs]: that optimization is unnecessary

File: src/libcollections/str.rs [hunk 2]
```diff
@@ -1683,6 +1683,49 @@ impl str {
     pub fn parse<F: FromStr>(&self) -> Result<F, F::Err> {
         core_str::StrExt::parse(self)
     }
+    
+    /// Returns a string slice extracted from the starting index to
+    /// the end of the string.
+    ///
+    /// Returns [`None`] if length is zero, or if string is empty.
```
Comment (oli-obk) [inline src/libcollections/str.rs]: why not simply return an empty string?

**Comment by oli-obk** [comment]
What's the motivation behind this change? What's the advantage over using the slicing syntax directly?

**Comment by RockyTV** [comment]
@oli-obk Using the slicing syntax directly can be confusing. For example, someone wanting to retrieve a substring (or, a slice) with length of 2 characters could think this would do it:
```
let string = "Hello, world!"
let substring = string[7, 2]; // we want to retrieve the 'wo' part
```

However, this gives a compile error, because the last parameter of the range syntax works like `start + desired length`. The above example should then be:
```
let substring = string[7, 9];
```

That's where I got the motivation to submit this Pull Request.

**Comment by oli-obk** [comment]
While the range syntax doesn't support this directly, the following works: `let substring = &string[7..][..2];` The first slice takes all characters starting at the 7th byte, and then the second slicing takes all the characters up to and including the 2nd byte from the first slice.

**Comment by bluss** [comment]
We do want some kind of Option-enabled slicing api, but the naming and api needs some adjustment. It should use a range as the argument, just like Index does, and like `String::drain()` does.

**Comment by RockyTV** [comment]
@bluss You're suggesting for me to remove the two methods and just go with one instead, that takes a range as an argument?

**Comment by bluss** [comment]
Yes. I don't know if we have or need an RFC for this, but it's in fact something we want. But not to have an alternative argument style, but to provide careful slicing (which returns None on out of bounds / indexing error).

**Comment by RockyTV** [comment]
As of now, as @bluss suggested, I decided to go with a `RangeArgument` as the only argument the function needs. Now, it will return None if no slice could be retrieved from the string.

**Comment by TimNN** [comment]
Looking at PR #31191 it seems as if moving `RangeArgument` to `core::ops` (and the using it throughout the standard library) is a little bit more involved than the two commits in this PR, so this PR may want to wait until #31191 has landed.

File: src/libcore/str/mod.rs [hunk 3]
```diff
@@ -1905,6 +1909,20 @@ impl StrExt for str {
...
+    #[inline]
+    fn substr<R>(&self, range: R) -> Option<&str> 
+        where R: RangeArgument<usize>
+    {
+        let len = self.len();
+        let start = *range.start().unwrap_or(&0);
+        let end = *range.end().unwrap_or(&len);
+        
+        if self.is_char_boundary(start) &&
+            self.is_char_boundary(end) { return Some(&self[start .. end]); }
```
Comment (bluss) [inline src/libcore/str/mod.rs]: This needs the condition `start <= end` too, to be correct (and never panic).

File: src/libcollections/str.rs [hunk 4]
```diff
@@ -1683,6 +1684,32 @@ impl str {
     pub fn parse<F: FromStr>(&self) -> Result<F, F::Err> {
         core_str::StrExt::parse(self)
     }
+    
+    /// Returns a string slice extracted from the starting index to
+    /// the specified length.
+    ///
+    /// Returns [`None`] if length is zero, or if string is empty.
+    /// # Example
```
Comment (bluss) [inline src/libcollections/str.rs]: Doc needs update for change of functionality. You could compare it with usual slicing (`&s[a..b]` style).

**Comment by RockyTV** [comment]
@TimNN I don't see how moving `RangeArgument` to `core::ops` is more than what I did. Doesn't #31191 allow indexing using `RangeArgument` instead of the special cases?

**Comment by TimNN** [comment]
Yeah, I looked over the PR's in more detail again and may habe been a bit hasty in coming to my conclusion when posting the previous comment. I mainly wanted create a link between both PR's (since they both moved `RangeArgument` to `core::ops`) and #31191 seemed relatively complex / involved whereas here the move seemed to not be the main focus of the PR but rather a later add-on.

**Comment by alexcrichton** [comment]
Thanks for the PR @RockyTV! This definitely seems related to https://github.com/rust-lang/rfcs/pull/1325 which @bluss mentioned in terms of an `Option`-returning slice API as well. I suspect that we're going to want to consider both `str` and slices at the same time when creating an API such as this.

Otherwise, it looks like (as others have commented here), this API is subsumed by `&foo[a..b]` today in terms of functionality (minus the return value of `Option`).

**Comment by alexcrichton** [comment]
Closing due to inactivity, but feel free to resubmit with a rebase!

**Pull request closed without merging**